### PR TITLE
fix(provisioner): emit URLs for all API addresses in FindTools

### LINKED ISSUE #22906: Azure: machines fail to provision due to tools URL only using first API address
## Description

When provisioning a new machine on Azure, the compute-provisioner worker constructs
download URLs for agent binaries using only the first API address (`addrs[0]`).
This is typically the cloud-local/private IP address. Machines in separate VNets
that cannot reach the private IP are stuck retrying the download forever and never
report ready.

Additionally, when the simplestreams fallback path is taken, raw simplestreams URLs
are returned unmodified — the URLs are never rewritten to point at the API server.

Both behaviours diverge from the server-side `FindAgents`→`ToolsURLs` logic, which
iterates **all** API addresses and produces one tools entry per address. The
cloud-init download template renders one `curl` per URL with sequential fallback.

## Juju version

4.1

## Cloud

Azure

## Expected behaviour

When adding a machine to a model on Azure, the machine should provision successfully
regardless of whether it is in the same VNet as the controller. The download script
should cycle through all API addresses (including the public one) and succeed.

## Reproduce / Test

1. Bootstrap a controller on Azure:
```
juju bootstrap azure
```

2. Add a model and a machine:
```
juju add-model test
juju add-machine
```

3. The machine never reports and stays in a pending state — the cloud-init tools
   download script has only the private address and cannot reach it from the new
   machine's VNet.

## Notes & References

- Discovered during work on #22758 (Azure dual-stack NSG firewaller)
- Fix: #22905

### PR DESCRIPTION
**Pull request by gfouillet**
## Why this change is needed and what it does

`toolsFinderAdapter.FindTools` in `internal/worker/computeprovisioner/provisioner_adapters.go`
uses `addrs[0]` (the first API address) when constructing download URLs for tools.
This means the cloud-init download script only gets a single URL — typically the
cloud-local/private address. Machines in other VNets that cannot reach the
private IP are stuck retrying forever.

When the simplestreams fallback path was taken, raw simplestreams URLs were
returned with no URL rewriting at all.

Both behaviours diverge from the server-side `FindAgents`→`ToolsURLs` logic,
which iterates **all** API addresses and produces one tools entry per address.
The cloud-init template renders one `curl` per URL with sequential fallback,
so a single URL means no fallback is possible.

This fix makes the adapter:

- Expand matched tools from agent-binary storage to one entry per API address
  (was `addrs[0]` only)
- Rewrite simplestreams-fallback tools to point at the API server and similarly
  expand them (was returning raw simplestreams URLs)
- Return an error when address lookup fails or returns no addresses instead of
  silently proceeding
- Safely copy shared pointers per entry to prevent aliasing
- Deduplicate the `coretools.Filter` construction (feedback from #22905 review)

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you are testing
- ~Integration tests~
- ~doc.go added or updated in changed packages~

## QA steps

This bug was discovered while working on #22758 — the Azure dual-stack NSG
firewaller PR. The warning block in that PR describes the issue; this PR is the
fix.

> [!WARNING]
> After merging, the warning block in #22758 should be updated to note that this
> PR fixes the connectivity issue described there.

1. Bootstrap a controller on Azure:
```sh
juju bootstrap azure
```

2. Add a model and a machine:
```sh
juju add-model test
juju add-machine
```

3. **Without the fix**: the machine never reports and stays in a pending state
   (the cloud-init tool download script has only the private address and cannot
   reach it from the new machine's VNet).

4. **With the fix**: the machine eventually starts — the download script cycles
   through all API addresses (including the public one) and succeeds.

## Follow-ups

- [#22910](https://github.com/juju/juju/issues/22910) — replace `fmt.Sprintf` URL construction with `url.URL` + `url.JoinPath` in both the worker adapter and the server-side `ToolsURL` helper.

## Links

**Issue:** Fixes #22906
**JIRA:** [JUJU-10167](https://warthogs.atlassian.net/browse/JUJU-10167)

[JUJU-10167]: https://warthogs.atlassian.net/browse/JUJU-10167?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by iyiguncevik — APPROVED** [review]

File: internal/worker/computeprovisioner/provisioner_adapters.go [hunk 1]
```diff
@@ -353,26 +361,52 @@ func (a *toolsFinderAdapter) FindTools(
...
+	if len(addrs) == 0 {
+		return nil, errors.New("no suitable API server address to pick from")
+	}
+
+	var fullList coretools.List
+	for _, baseTools := range baseList {
+		for _, addr := range addrs {
+			tools := *baseTools // copy — don't mutate the shared pointer
+			serverRoot := fmt.Sprintf("https://%s/model/%s", addr, a.modelUUID)
+			tools.URL = fmt.Sprintf("%s/tools/%s", serverRoot, tools.Version.String())
```
Comment (SimonRichardson) [inline internal/worker/computeprovisioner/provisioner_adapters.go:402]: We should use the `url.URL` and `url.JoinPath` instead of constructing the URL.

**Review by SimonRichardson — APPROVED** [review]

**Comment by gfouillet** [comment]
## Review responses

**@SimonRichardson** ([#discussion_r3594172186](https://github.com/juju/juju/pull/22905#discussion_r3594172186)):
Acknowledged. Opened [#22910](https://github.com/juju/juju/issues/22910) tracking `url.URL` + `url.JoinPath` refactoring in both the worker adapter (`provisioner_adapters.go:400-401`) and the server-side `ToolsURL` helper (`apiserver/common/tools.go:409-419`). An inline `TODO` has been added above the current `fmt.Sprintf` lines.

**@lee-van-cleef** (auto-review):
- ✅ **Missing test** — added `TestFindToolsListAgentBinariesErrorFallsBackToSimplestreams`, covering the `listErr != nil` + simplestreams-success path (suggestion 1).
- ✅ **Filter deduplication** — hoisted `coretools.Filter{Number, OSType, Arch}` to the top of `FindTools` and reuse it in both the storage and simplestreams branches (suggestion 3).
- ℹ️ **Simplestreams URL overwrite** (suggestion 2) — left as-is; the pre-existing assumption is already encoded in the server-side `ToolsURL` helper and tracked in [#22910](https://github.com/juju/juju/issues/22910). This is not a new behaviour introduced by this PR.
